File: android/src/main/java/com/rndocumentscanner/utils/ImageValidator.kt
package com.rndocumentscanner.utils

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.media.FaceDetector
import android.net.Uri
import com.margelo.nitro.NitroModules
import java.io.File

object ImageValidator {

    data class ValidationResult(val hasFace: Boolean, val hash: String)

    /**
     * Kiểm tra ảnh: phát hiện khuôn mặt và sinh mã băm dHash (64-bit hex)
     */
    fun validate(imageUri: String): ValidationResult {
        val bitmap = loadImage(imageUri) ?: return ValidationResult(false, "")
        try {
            val hasFace = detectFace(bitmap)
            val hash = computeDHash(bitmap)
            return ValidationResult(hasFace, hash)
        } finally {
            bitmap.recycle()
        }
    }

    /**
     * So sánh độ tương đồng giữa 2 ảnh (0.0 đến 1.0) dựa trên khoảng cách Hamming của dHash
     */
    fun compare(imageUri1: String, imageUri2: String): Double {
        val b1 = loadImage(imageUri1)
        val b2 = loadImage(imageUri2)
        if (b1 == null || b2 == null) {
            b1?.recycle()
            b2?.recycle()
            return 0.0
        }

        try {
            val hash1 = computeDHash(b1)
            val hash2 = computeDHash(b2)
            return similarity(hash1, hash2)
        } finally {
            b1.recycle()
            b2.recycle()
        }
    }

    // MARK: - Private Helpers

    private fun loadImage(uriString: String): Bitmap? {
        val clean = if (uriString.startsWith("file://")) uriString.substring(7) else uriString
        val file = File(clean)
        if (file.exists()) {
            return BitmapFactory.decodeFile(file.absolutePath)
        }
        try {
            val uri = Uri.parse(uriString)
            val context = NitroModules.applicationContext
            if (context != null && uri.scheme == "content") {
                context.contentResolver.openInputStream(uri)?.use { stream ->
                    return BitmapFactory.decodeStream(stream)
                }
            }
        } catch (_: Exception) {}
        return null
    }

    /**
     * Phát hiện khuôn mặt bằng android.media.FaceDetector tiêu chuẩn của Android SDK
     */
    private fun detectFace(bitmap: Bitmap): Boolean {
        var scaled: Bitmap? = null
        var rgb565: Bitmap? = null
        try {
            val targetWidth = 400
            val targetHeight = (bitmap.height.toFloat() / bitmap.width.toFloat() * targetWidth).toInt()
            val evenWidth = if (targetWidth % 2 == 0) targetWidth else targetWidth - 1
            val evenHeight = if (targetHeight % 2 == 0) targetHeight else targetHeight - 1

            scaled = Bitmap.createScaledBitmap(bitmap, evenWidth, evenHeight, true)
            rgb565 = scaled.copy(Bitmap.Config.RGB_565, false) ?: return false

            val detector = FaceDetector(evenWidth, evenHeight, 1)
            val faces = arrayOfNulls<FaceDetector.Face>(1)
            val count = detector.findFaces(rgb565, faces)
            return count > 0 && faces[0] != null && faces[0]!!.confidence() > 0.4f
        } catch (_: Exception) {
            return false
        } finally {
            if (scaled != null && scaled != bitmap) scaled.recycle()
            if (rgb565 != null && rgb565 != bitmap) rgb565.recycle()
        }
    }

    /**
     * Tính mã băm Difference Hash (dHash) 64-bit
     */
    private fun computeDHash(bitmap: Bitmap): String {
        var scaled: Bitmap? = null
        try {
            scaled = Bitmap.createScaledBitmap(bitmap, 9, 8, true)
            var hash = 0L
            for (y in 0 until 8) {
                for (x in 0 until 8) {
                    val leftPixel = scaled.getPixel(x, y)
                    val rightPixel = scaled.getPixel(x + 1, y)

                    val leftGray = (Color.red(leftPixel) * 299 +
                                    Color.green(leftPixel) * 587 +
                                    Color.blue(leftPixel) * 114) / 1000
                    val rightGray = (Color.red(rightPixel) * 299 +
                                     Color.green(rightPixel) * 587 +
                                     Color.blue(rightPixel) * 114) / 1000

                    hash = (hash shl 1) or (if (leftGray > rightGray) 1L else 0L)
                }
            }
            return String.format("%016x", hash)
        } catch (_: Exception) {
            return ""
        } finally {
            if (scaled != null && scaled != bitmap) scaled.recycle()
        }
    }

    /**
     * Tính độ tương đồng Hamming từ 2 mã băm Hex 64-bit
     */
    private fun similarity(hash1: String, hash2: String): Double {
        if (hash1.isEmpty() || hash2.isEmpty() || hash1.length != 16 || hash2.length != 16) {
            return 0.0
        }
        return try {
            val val1 = java.lang.Long.parseUnsignedLong(hash1, 16)
            val val2 = java.lang.Long.parseUnsignedLong(hash2, 16)
            val xorVal = val1 xor val2
            val diffBits = java.lang.Long.bitCount(xorVal)
            maxOf(0.0, 1.0 - (diffBits.toDouble() / 64.0))
        } catch (_: Exception) {
            0.0
        }
    }
}

File: ios/Utils/ImageValidator.swift
import UIKit
import Vision
import CoreImage

class ImageValidator {

    /// Kiểm tra ảnh: phát hiện khuôn mặt và sinh mã băm dHash (64-bit hex)
    static func validate(imageUri: String) -> (hasFace: Bool, hash: String) {
        return autoreleasepool {
            guard let image = loadImage(from: imageUri), let cgImage = image.cgImage else {
                return (false, "")
            }

            let hasFace = detectFace(cgImage: cgImage)
            let hash = computeDHash(image: image)

            return (hasFace, hash)
        }
    }

    /// So sánh độ tương đồng giữa 2 ảnh (0.0 đến 1.0) dựa trên khoảng cách Hamming của dHash
    static func compare(imageUri1: String, imageUri2: String) -> Double {
        return autoreleasepool {
            guard let img1 = loadImage(from: imageUri1),
                  let img2 = loadImage(from: imageUri2) else {
                return 0.0
            }

            let hash1 = computeDHash(image: img1)
            let hash2 = computeDHash(image: img2)

            return similarity(hash1: hash1, hash2: hash2)
        }
    }

    // MARK: - Private Helpers

    private static func loadImage(from uri: String) -> UIImage? {
        let cleanUri = uri.replacingOccurrences(of: "file://", with: "")
        if let image = UIImage(contentsOfFile: cleanUri) {
            return image
        }
        if let url = URL(string: uri), let data = try? Data(contentsOf: url) {
            return UIImage(data: data)
        }
        return nil
    }

    /// Phát hiện khuôn mặt bằng Vision Framework
    private static func detectFace(cgImage: CGImage) -> Bool {
        var foundFace = false
        let request = VNDetectFaceRectanglesRequest { req, err in
            guard err == nil, let results = req.results as? [VNFaceObservation] else { return }
            foundFace = !results.isEmpty
        }

        let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
        try? handler.perform([request])
        return foundFace
    }

    /// Tính mã băm Difference Hash (dHash) 64-bit
    private static func computeDHash(image: UIImage) -> String {
        let targetWidth = 9
        let targetHeight = 8

        // Vẽ ảnh về kích thước 9x8 grayscale
        let colorSpace = CGColorSpaceCreateDeviceGray()
        guard let context = CGContext(
            data: nil,
            width: targetWidth,
            height: targetHeight,
            bitsPerComponent: 8,
            bytesPerRow: targetWidth,
            space: colorSpace,
            bitmapInfo: CGImageAlphaInfo.none.rawValue
        ), let cgImage = image.cgImage else {
            return ""
        }

        context.draw(cgImage, in: CGRect(x: 0, y: 0, width: targetWidth, height: targetHeight))
        guard let pixelData = context.data else { return "" }

        let buffer = pixelData.bindMemory(to: UInt8.self, capacity: targetWidth * targetHeight)

        var hash: UInt64 = 0
        for y in 0..<targetHeight {
            for x in 0..<(targetWidth - 1) {
                let leftPixel = buffer[y * targetWidth + x]
                let rightPixel = buffer[y * targetWidth + (x + 1)]
                hash = (hash << 1) | (leftPixel > rightPixel ? 1 : 0)
            }
        }

        return String(format: "%016llx", hash)
    }

    /// Tính độ tương đồng Hamming từ 2 mã băm Hex 64-bit
    private static func similarity(hash1: String, hash2: String) -> Double {
        guard let val1 = UInt64(hash1, radix: 16),
              let val2 = UInt64(hash2, radix: 16) else {
            return 0.0
        }

        let xorVal = val1 ^ val2
        let differingBits = xorVal.nonzeroBitCount
        return max(0.0, 1.0 - (Double(differingBits) / 64.0))
    }
}
